ports/esp32/boards/TINYPICO/modules/tinypico.py

```diff
@@ -0,0 +1,136 @@
+    """
+    adc = ADC( Pin( BAT_VOLTAGE ) )         # Assign the ADC pin to read
+    measuredvbat = adc.read()               # Read the value
+    measuredvbat /= 4095                    # divide by 4095 as we are using the default ADC voltage range of 0-1V
+    measuredvbat *= 3.7                     # Multiply by 3.7V, our reference voltage
+    return measuredvbat
+
+# Return the current charge state of the battery - we need to read the value multiple times
+# to eliminate false negatives due to the charge IC not knowing the difference between no battery
+# and a full battery not charging - This is why the charge LED flashes
+def get_battery_charging():
+    """
+    Returns the current battery charging state.
+    This can trigger false positives as the charge IC can't tell the difference between a full battery or no battery connected.
+    """
+    measuredVal = 0                         # start our reading at 0
+    io = Pin( BAT_CHARGE, Pin.IN )          # Assign the pin to read
+  
+    for y in range(0, 10):                  # loop through 10 times adding the read values together to ensure no false positives    
+        measuredVal += io.value()
+
+    return ( measuredVal == 0 )             # return True if the value is 0
+
+# Return the internal PICO-D4 temperature in Farenheit
+def get_internal_temp_F():
+    """Return the internal PICO-D4 temperature in farenheit."""
+    return esp32.raw_temperature()
+
+# Return the internal PICO-D4 temperature in Celcius
+def get_internal_temp_C():
+    """Return the internal PICO-D4 temperature in celcius."""
+    return ( esp32.raw_temperature() - 32 ) / 1.8
+
+# Power to the on-oard Dotstar is controlled by a PNP transistor, so low is ON and high is OFF
+# We also need to set the Dotstar clock and data pins to be inputs to prevent power leakage when power is off
+# This might be improved at a future date
+# The reason we have power control for the Dotstar is that it has a quiescent current of around 1mA, so we
+# need to be able to cut power to it to minimise power consumption during deep sleep or with general battery powered use
+# to minimse un-needed battery drain
+def set_dotstar_power( state ):
+    """Set the power for the on-board Dostar to allow no current draw when not needed."""
+    # Set the power pin to the inverse of state 
+    if state:
+        Pin( DOTSTAR_PWR, Pin.OUT, None )   # Break the PULL_HOLD on the pin
+        Pin( DOTSTAR_PWR ).value( False )   # Set the pin to LOW to enable the Transistor
+    else:
+        Pin(13, Pin.IN, Pin.PULL_HOLD) # Set PULL_HOLD on the pin to allow the 3V3 pullup to work
+                   
+    Pin(DOTSTAR_CLK, Pin.OUT if state else Pin.IN )     # If power is on, set CLK to be output, otherwise input
+    Pin(DOTSTAR_DATA, Pin.OUT if state else Pin.IN )    # If power is on, set DATA to be output, otherwise input
+
+    # A small delay to let the IO change state
+    time.sleep(.035)
```

Comment:
Did you observe a need for this in testing? I'd have thought the GPIO would change start faster than Python could return to the next opcode. Interesting!

Reply:
Yes, it's needed unfortunately. It's ugly :(

Reply:
Ew :( my condolences.